plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -456,26 +458,29 @@ public DeltaLakeMetadata(
         this.allowManagedTableRename = allowManagedTableRename;
     }
 
-    public TableSnapshot getSnapshot(ConnectorSession session, SchemaTableName table, String tableLocation, long atVersion)
-    {
-        return getSnapshot(session, table, tableLocation, Optional.of(atVersion));
-    }
-
-    @VisibleForTesting
-    protected TableSnapshot getSnapshot(ConnectorSession session, SchemaTableName table, String tableLocation, Optional<Long> atVersion)
+    public TableSnapshot getSnapshot(
+            ConnectorSession session,
+            SchemaTableName table,
+            String tableLocation,
+            Optional<Long> atVersion,
+            Optional<ConnectorTableVersion> endVersion)
     {
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        Optional<Long> endSnapshotVersion = endVersion.map(version -> getSnapshotIdFromVersion(fileSystem, tableLocation, version));
         try {
             if (atVersion.isEmpty()) {
                 atVersion = Optional.ofNullable(queriedVersions.get(table));
             }
             if (atVersion.isPresent()) {
                 long version = atVersion.get();
                 TableSnapshot snapshot = queriedSnapshots.get(new QueriedTable(table, version));
-                checkState(snapshot != null, "No previously loaded snapshot found for query %s, table %s [%s] at version %s", session.getQueryId(), table, tableLocation, version);
+                if (snapshot == null) { // This may happen when INSERT statement uses time travel in the source table
+                    return transactionLogAccess.loadSnapshot(session, table, tableLocation, Optional.of(version));
+                }
                 return snapshot;
             }
 
-            TableSnapshot snapshot = transactionLogAccess.loadSnapshot(session, table, tableLocation);
+            TableSnapshot snapshot = transactionLogAccess.loadSnapshot(session, table, tableLocation, endSnapshotVersion);
             // Lack of concurrency for given query is currently guaranteed by DeltaLakeMetadata
             checkState(queriedVersions.put(table, snapshot.getVersion()) == null, "queriedLocations changed concurrently for %s", table);
             queriedSnapshots.put(new QueriedTable(table, snapshot.getVersion()), snapshot);
```

Comment:
Is it guaranteed that `getSnapshot` will be executed for targetTable first and later for source tables?
In case of the same table with different version in source and target we will end up in caching only one of them. Should it always be target or it does not matter?